src/common/test-json.c

```diff
@@ -676,6 +676,155 @@ test_write_infinite_nan (void)
   g_free (string);
 }
 
+static JsonNode *
+flip_integer (JsonNode *node,
+              gpointer  user_data)
+{
+  int *counter = user_data;
+
+  if (json_node_get_value_type (node) != G_TYPE_INT64)
+    return NULL;
+
+  int value = json_node_get_int (node);
+
+  (*counter)++;
+
+  return json_node_init_int (json_node_alloc(), -value);
+}
+
+/* replace all three-character strings with the word 'three' */
+static JsonNode *
+three_subst (JsonNode *node,
+             gpointer  user_data)
+{
+  int *counter = user_data;
+
+  const gchar *value = json_node_get_string (node);
+
+  if (value == NULL || strlen (value) != 3)
+    return NULL;
+
+  (*counter)++;
+
+  return json_node_init_string (json_node_alloc(), "three");
+}
+
+static JsonNode *
+null_subst (JsonNode *node,
+            gpointer  user_data)
+{
+  int *counter = user_data;
+
+  (*counter)++;
+
+  return NULL;
+}
+
+static void
+test_walk (void)
+{
+  GError *error = NULL;
+
+  g_autoptr(JsonObject) source = cockpit_json_parse_object (patch_data, -1, &error);
+  g_assert_no_error (error);
+  json_object_seal (source);
+
+  /* for the nul case, we should get exactly the same object back */
+  {
+    int counter = 0;
+    g_autoptr(JsonObject) result = cockpit_json_walk (source, null_subst, &counter);
+    g_assert_cmpint (counter, ==, 13);
```

Comment:
Meh magic numbers -- but I suppose there's no easy way to get the node count of a json tree, at least I didn't see any in the docs.